src/vm/eventpipebuffermanager.cpp

```diff
@@ -960,4 +881,44 @@ bool EventPipeBufferList::EnsureConsistency()
 }
 #endif // _DEBUG
 
+// This struct is a TLS wrapper around a pointer to thread-specific EventPipeBufferList
+// The struct wrapper is present mainly because we need a way to free the EventPipeBufferList
+// when the thread that owns it dies. Placing this struct as a TLS variable will call ~ThreadEventBufferList()
+// when the thread dies so we can free EventPipeBufferList in the destructor.  
+struct ThreadEventBufferList
+{
+    EventPipeBufferList* m_pThreadEventBufferList;
+
+    ~ThreadEventBufferList()
+    {
+        // Before the thread dies, mark its buffers as no longer owned
+        // so that they can be cleaned up after the thread dies.
+        EventPipeBufferList *pList = GetThreadEventBufferList();
+        if (pList != NULL)
+            pList->SetOwnedByThread(false);
+    }
+};
+
+
+extern "C" {
+#ifndef __llvm__
+__declspec(thread) ThreadEventBufferList gCurrentThreadEventBufferList = { NULL };
+#else // !__llvm__
+thread_local ThreadEventBufferList gCurrentThreadEventBufferList = { NULL };
+#endif // !__llvm__
+} // extern "C"
+
+
+EXTERN_C inline EventPipeBufferList* STDCALL GetThreadEventBufferList()
```

Comment:
These do not need to be `EXTERN_C`. In fact, these can be a static methods inside EventPipeBufferList.

Reply:
(or inside ThreadEventBufferList)